happi/tests/test_cli.py

```diff
@@ -95,9 +122,200 @@ def test_search_z(happi_cfg):
     assert [r.device for r in res] == [r.device for r in res_cli]
 
 
-def test_search_z_range(happi_cfg):
+def test_search_z_range(happi_cfg, caplog):
     client = happi.client.Client.from_config(cfg=happi_cfg)
     res = client.search_range('z', 3.0, 6.0)
     res_cli = happi.cli.happi_cli(['--verbose', '--path', happi_cfg, 'search',
                                    'z=3.0,6.0'])
     assert [r.device for r in res] == [r.device for r in res_cli]
+    # test invalid range
+    with caplog.at_level(logging.ERROR):
+        res_cli = happi.cli.happi_cli(['--verbose', '--path', happi_cfg,
+                                       'search', 'z=6.0,3.0'])
+        assert "Invalid range, make sure start < stop" in caplog.text
+        caplog.clear()
+
+
+def test_both_range_and_regex_search(happi_cfg):
+    client = happi.client.Client.from_config(cfg=happi_cfg)
+    # we're only interested in getting this entry (TST_BASE_PIM2)
+    res = client.search_regex(z='6.0')
+    # we're going to search for z=3,7 name=TST_BASE_PIM2
+    # we should only get in return one entry, even though the z value found 2
+    res_cli = happi.cli.happi_cli(['--verbose', '--path', happi_cfg, 'search',
+                                   'name=TST_BASE_PIM2', 'z=3.0,7.0'])
+    assert [r.device for r in res] == [r.device for r in res_cli]
+
+
+@pytest.mark.parametrize("from_user, expected_output", [
+    # Test add item - succeeding
+    pytest.param(['HappiItem', 'happi_name', 'device_class', ['arg1', 'arg2'],
+                  {'name': 'my_name'}, True, 'docs', 'y'], [
+        "Please select a container, or press enter for generic "
+        "Ophyd Device container",
+        "Enter value for name, default=None, "
+        "enforce=re.compile('[a-z][a-z\\\\_0-9]{2,78}$')",
+        "Enter value for device_class, default=None, enforce=<class 'str'>",
+        "Enter value for args, default=[], enforce=<class 'list'>",
+        "Enter value for kwargs, default={}, enforce=<class 'dict'>",
+        "Enter value for active, default=True, enforce=<class 'bool'>",
+        "Enter value for documentation, default=None, enforce=<class 'str'>",
+        "Please confirm the following info is correct:",
+        "Adding device",
+        "Storing device HappiItem (name=happi_name) ...",
+        "Adding / Modifying information for happi_name ...",
+        "HappiItem HappiItem (name=happi_name) has been "
+        "succesfully added to the database"],
+    ),
+    # Test add item - aborting
+    pytest.param(['HappiItem', 'happi_name2', 'device_class', ['arg1', 'arg2'],
```

Comment:
Good on the `pytest.param`, but you're still missing the `id='useful_identifier'` keyword argument.

Reply:
ohh i see hahah, let me add that in

Reply:
Ok i think this is what you meant: 1568fc7 - but please let me know if I still did not get it 😄 - i won't mind, i don't think I have ever used the `ids` before  - it's good to know about them.
Also, please let me know if I should have provided something more generic for the IDs. 

I am able to now use the `-k id_value` to test a specific one which is cool